### DISCUSSION TITLE
Conversation about better Golang query API

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#2065] [Q&A]
Lately I've been using a wrapper for saved queries in order to get around the terrible ergonomics of a normal record not found returning an error of `sql.ErrNoRows`.  The problem is first needing to check if the error is specifically `sql.ErrNoRows`, then checking for if it's an error.  It makes ugly code because I typically want to just check for it being an error first in order to return.  But I can't do that since it could be `sql.ErrNoRows`.

I created two flavors of query wrappers.  One makes the `model == nil` mean that it wasn't found.  The other adds an extra variable called found.

Does anyone have a suggestion, or is there something that I've missed?

```golang
func (c *ctrl) getHome(ctx echo.Context) error {
	u := auth.User(ctx)
	if u == nil {
		return ctx.NoContent(http.StatusInternalServerError)
	}

	m, err := models.SiteByUserId(c.Dao(), u.Id)
	if err != nil {
		return ctx.NoContent(http.StatusInternalServerError)
	}
	if m == nil {
		return ctx.Redirect(http.StatusSeeOther, "/site/create")
	}

	return ctx.Render(http.StatusOK, "home/index", nil)
}
```

```golang
func SiteByUserId(dao *daos.Dao, id string) (*models.Record, error) {
	return query.UsingNilRecord(
		dao.FindFirstRecordByData(Sites, "user", id),
	)
}
```

```golang
import (
	"database/sql"

	"github.com/pocketbase/pocketbase/models"
)

func UsingNilRecord(m *models.Record, err error) (*models.Record, error) {
	var finalModel *models.Record
	if err == nil {
		finalModel = m
	}

	var finalErr error
	if err != nil && err != sql.ErrNoRows {
		finalErr = err
	}

	return finalModel, finalErr
}

func UsingFound(m *models.Record, err error) (*models.Record, bool, error) {
	found := m != nil && err == nil

	var finalErr error
	if err != nil && err != sql.ErrNoRows {
		finalErr = err
	}

	return m, found, finalErr
}
```

**Comment by ganigeorgiev** [comment]
I'm not sure that I understand what the wrappers are supposed to do.

In case of an error, the pointer value (`*models.Record`) will be always `nil`.

If you want to distinguish a "Not found" from other db errors, then you have to check it against `sql.ErrNoRows` (either directly as in your examples or via `errors.Is(err, sql.ErrNoRows)`).

**Reply by mlusetti** [reply]
The main point here seems to be the "empty result set" treated as an error while it is not an error in the first place.

At least this is what I understand from the original post.

**Reply by dschissler** [reply]
@ganigeorgiev I want to be able to check in the order

1) If (actual) error
2) If not found
3) If found (or without a check since it's the last case and to not have indentation.)

The problem is that it's needed to be

1) If error and is not notFoundErr
2) If not found
3) If found.

By default the ergonomics are simply poor.  I don't think that there is any reason for a query in a form or a route to need to reference `sql.ErrNoRows` because that can be smoothed out via the helper wrappers that I provided.  Then that provides for methods which make sense to a human.
